src/Language/PureScript/ModuleDependencies.hs

```diff
@@ -1,99 +1,65 @@
+sortModules
+  :: forall m
+   . MonadError MultipleErrors m
+  => [Module]
+  -> m ([Module], ModuleGraph)
 sortModules ms = do
-  let verts = map goModule ms
-  ms' <- mapM toModule $ stronglyConnComp verts
-  let (graph, fromVertex, toVertex) = graphFromEdges verts
-      moduleGraph = do (_, mn, _) <- verts
-                       let v       = fromMaybe (internalError "sortModules: vertex not found") (toVertex mn)
-                           deps    = reachable graph v
-                           toKey i = case fromVertex i of (_, key, _) -> key
-                       return (mn, filter (/= mn) (map toKey deps))
-  return (ms', moduleGraph)
+    let mns = map getModuleName ms
+    verts <- mapM (toGraphNode mns) ms
+    ms' <- mapM toModule $ stronglyConnComp verts
+    let (graph, fromVertex, toVertex) = graphFromEdges verts
+        moduleGraph = do (_, mn, _) <- verts
+                         let v       = fromMaybe (internalError "sortModules: vertex not found") (toVertex mn)
+                             deps    = reachable graph v
+                             toKey i = case fromVertex i of (_, key, _) -> key
+                         return (mn, filter (/= mn) (map toKey deps))
+    return (ms', moduleGraph)
   where
-  goModule :: Module -> (Module, ModuleName, [ModuleName])
-  goModule m@(Module _ _ _ ds _) =
-    let ams = concatMap extractQualAs ds
-    in (m, getModuleName m, nub (concatMap (usedModules ams) ds))
+    toGraphNode :: [ModuleName] -> Module -> m (Module, ModuleName, [ModuleName])
+    toGraphNode mns m@(Module _ _ mn ds _) = do
+      let deps = nub (concatMap usedModules ds)
+      forM_ deps $ \dep ->
+        when (dep `notElem` mns) $
+          throwError . addHint (ErrorInModule mn) . errorMessage $ ModuleNotFound dep
+      pure (m, getModuleName m, deps)
 
-  -- Extract module names that have been brought into scope by an `as` import.
-  extractQualAs :: Declaration -> [ModuleName]
-  extractQualAs (PositionedDeclaration _ _ d) = extractQualAs d
-  extractQualAs (ImportDeclaration _ _ (Just am)) = [am]
-  extractQualAs _ = []
-
--- |
--- Calculate a list of used modules based on explicit imports and qualified
--- names. `ams` is a list of `ModuleNames` that refer to names brought into
--- scope by importing with `as` - this ensures that when building the list we
--- don't inadvertantly assume a dependency on an actual module, if there is a
--- module that has the same name as the qualified import.
---
-usedModules :: [ModuleName] -> Declaration -> [ModuleName]
-usedModules ams d =
-  let (f, _, _, _, _) = everythingOnValues (++) forDecls forValues (const []) (const []) (const [])
-      (g, _, _, _, _) = accumTypes (everythingOnTypes (++) forTypes)
-  in nub (f d ++ g d)
-  where
+-- | Calculate a list of used modules based on explicit imports and qualified names.
+usedModules :: Declaration -> [ModuleName]
+usedModules d = nub (f d) where
```

Comment:
We could just use a `Set` here anyway perhaps? I think there are a few places we use lists where sets would better express the intention and probably perform better also.